src/SignalR/common/Shared/ClientResultsManager.cs

```diff
@@ -0,0 +1,198 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.AspNetCore.SignalR.Protocol;
+
+namespace Microsoft.AspNetCore.SignalR.Internal;
+
+// Common type used by our HubLifetimeManager implementations to manage client results.
+// Handles cancellation, cleanup, and completion, so any bugs or improvements can be made in a single place
+internal class ClientResultsManager : IInvocationBinder
+{
+    private readonly ConcurrentDictionary<string, (Type Type, string ConnectionId, Func<CompletionMessage, Task> Completion)> _pendingInvocations = new();
+
+    public Task<T> AddInvocation<T>(string connectionId, string invocationId, CancellationToken cancellationToken)
+    {
+        var tcs = new TaskCompletionSourceWithCancellation<T>(this, connectionId, invocationId, cancellationToken);
+        _pendingInvocations.TryAdd(invocationId, (typeof(T), connectionId, completionMessage =>
+        {
+            if (completionMessage.HasResult)
+            {
+                tcs.SetResult((T)completionMessage.Result);
+            }
+            else
+            {
+                tcs.SetException(new Exception(completionMessage.Error));
+            }
+            return Task.CompletedTask;
+        }
+        ));
+
+        return tcs.Task;
+    }
+
+    public void AddInvocation(string invocationId, (Type Type, string ConnectionId, Func<CompletionMessage, Task> Completion) invocationInfo)
+    {
+        _pendingInvocations.TryAdd(invocationId, invocationInfo);
+    }
+
+    public Task TryCompleteResult(string connectionId, CompletionMessage message)
+    {
+        if (_pendingInvocations.TryGetValue(message.InvocationId!, out var item))
+        {
+            if (item.ConnectionId != connectionId)
+            {
+                throw new Exception("wrong ID");
+            }
+
+            // if false the connection disconnected right after the above TryGetValue
+            // or someone else completed the invocation (likely a bad client)
+            // we'll ignore both cases
+            if (_pendingInvocations.Remove(message.InvocationId!, out _))
+            {
+                return item.Completion(message);
+            }
+        }
+        else
+        {
+            // connection was disconnected or someone else completed the invocation
+        }
+        return Task.CompletedTask;
+    }
+
+    public (Type Type, string ConnectionId, Func<CompletionMessage, Task> Completion) RemoveInvocation(string invocationId)
+    {
+        _pendingInvocations.Remove(invocationId, out var item);
```

Comment:
TryRemove?
```suggestion
        _pendingInvocations.TryRemove(invocationId, out var item);
```

Reply:
Remove and TryRemove are the same thing in ConcurrentDictionary, do you prefer TryRemove?

Reply:
Yah